### PR TITLE
Implement a `explicit_generic_args_with_impl_trait` feature gate

### PR DESCRIPTION
**Pull request by nbdd0121**
Implements #83701

When this gate is enabled, explicit generic arguments can be specified even if `impl Trait` is used in argument position. Generic arguments can only be specified for explicit generic parameters but not for the synthetic type parameters from  `impl Trait`

So code like this will be accepted:
```rust
#![feature(explicit_generic_args_with_impl_trait)]

fn foo<T: ?Sized>(_f: impl AsRef<T>) {}
fn main() {
    foo::<str>("".to_string());
}
```

### CODE REVIEW

**Comment by bjorn3** [comment]
I have no experience with the typechecker, maybe

r? @lcnr

(another github suggestion)

**Comment by jackh726** [comment]
I will take this review if @lcnr can't.

That being said, I think it's prudent to get a sign off from the lang team here. I imagine this probably falls under the "make more reversible decisions" part of the initiative process. So probably just need *someone* to give the okay. So cc @rust-lang/lang

**Comment by joshtriplett** [comment]
I agree---since this is nightly-only, I think it's fine to go ahead with it.

I'll nominate this for consideration at a future meeting, but in the interim, go ahead.

File: compiler/rustc_feature/src/active.rs
Comment (jackh726) [inline compiler/rustc_feature/src/active.rs:691]: I suppose we can just co-opt that issue to be a tracking issue, yeah.

File: compiler/rustc_typeck/src/astconv/generics.rs
Comment (jackh726) [inline compiler/rustc_typeck/src/astconv/generics.rs:462]: I'd like a small comment here just explaining what we're doing.

File: src/test/ui/impl-trait/explicit-generic-args-for-impl.rs
Comment (jackh726) [inline src/test/ui/impl-trait/explicit-generic-args-for-impl.rs:1]: Can we put these tests in a directory? `src/test/ui/impl-trait/explicit-generic-args-with-impl-trait`

File: src/test/ui/impl-trait/explicit-generic-args-for-impl.stderr
Comment (jackh726) [inline src/test/ui/impl-trait/explicit-generic-args-for-impl.stderr:9]: Doesn't need to block this PR, but we should note in the tracking issue that there's an opportunity for better diagnostics to explicitly call out that you can't assign generics to `impl Trait`

File: src/test/ui/impl-trait/feature-gate-explicit_generic_args_with_impl_trait.stderr
Comment (jackh726) [inline src/test/ui/impl-trait/feature-gate-explicit_generic_args_with_impl_trait.stderr:1]: ...in fact, I imagine the longer for that part of the diagnostic will probably co-opt some of this logic

**Review by jackh726 — APPROVED** [review]
A couple nits, but r=me for this after those are addressed.

#83701 needs modified to "become" a tracking issue before/when this lands (or a new issue could be opened)

File: src/test/ui/impl-trait/explicit-generic-args-for-impl.rs
Comment (nbdd0121) [reply src/test/ui/impl-trait/explicit-generic-args-for-impl.rs:1]: What name should it called when it's moved into a directory? I am not at good at naming things :D

Comment (jackh726) [reply src/test/ui/impl-trait/explicit-generic-args-for-impl.rs:1]: Test names can be the same. Just allows us to group them better.

File: src/test/ui/impl-trait/explicit-generic-args-for-impl.stderr
Comment (nbdd0121) [reply src/test/ui/impl-trait/explicit-generic-args-for-impl.stderr:9]: I actually think from a pure syntax point of view it's natural to have this error, because indeed only 1 generic parameters are defined.

Comment (jackh726) [reply src/test/ui/impl-trait/explicit-generic-args-for-impl.stderr:9]: Oh, the error as-is is *fine*. But it couldn't hurt to have a "note: you aren't allowed to explicitly specify `impl Trait` args as generics" or something of the sort. This is especially true if people assume they can just replace generics with `impl Trait` and then get confused when they can no longer specify generics.

File: compiler/rustc_typeck/src/astconv/generics.rs
Comment (nbdd0121) [reply compiler/rustc_typeck/src/astconv/generics.rs:466]: Done

**Pull request merged**
